### PR TITLE
CNS-531: Save rewards for delegators in relay payment

### PR DESCRIPTION
**Pull request by oren-lava**
In this PR I've modified the provider reward when executing a relay payment TX so it'll consider delegations.
I've also created a delegator reward map that keeps a rewards record until they claim their rewards.

### CODE REVIEW

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 1]
```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+	if !found {
+		return math.Int{}, utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	// sanity check
+	if stakeEntry.DelegateCommission > 100 {
```
Comment (Yaroms) [inline x/pairing/keeper/msg_server_relay_payment.go:437]: this is checked when we stake, should we really check it here?
if we do, lets use 100 and and print an error

Comment (Yaroms) [reply x/pairing/keeper/msg_server_relay_payment.go:437]: also feels like this method is redundent, only fetches the stakeentry and calls another function.
lets merge them?

File: x/dualstaking/keeper/delegator_reward.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// CalcDelegatorsReward calculates the total amount of rewards for all delegators
+// delegatorsReward = totalReward - providerReward
+func (k Keeper) CalcDelegatorsReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
+	return totalReward.Sub(k.CalcProviderReward(stakeEntry, totalReward))
+}
+
+// CalcEffectiveDelegationsAndStake calculates the effective stake and effective delegations (for delegator rewards calculations)
+// effectiveDelegations = totalDelegations
+// effective stake = effectiveDelegations + providerStake
+func (k Keeper) CalcEffectiveDelegationsAndStake(stakeEntry epochstoragetypes.StakeEntry) (effectiveDelegations math.Int, effectiveStake math.Int) {
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegator_reward.go:85]: lets call them with correct naming,
effectiveDelegations -> Delegations
effectiveStake -> TotalDelegations

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 3]
```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+	err = k.updateDelegatorsReward(ctx, providerAddr, delegations, chainID, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot update delegators reward map", err)
+	}
+
+	return nil
+}
+
+// calculateProviderDelegatorsRewards returns the provider reward amount and updates the delegator reward map
+func (k Keeper) calculateProviderDelegatorsRewards(ctx sdk.Context, provider sdk.AccAddress, chainID string, totalReward math.Int) (math.Int, error) {
```
Comment (Yaroms) [inline x/pairing/keeper/msg_server_relay_payment.go:426]: this method does not update anything, update the comment

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 4]
```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "commission", Value: stakeEntry.DelegateCommission},
+		)
+	}
+
+	return k.dualStakingKeeper.CalcProviderReward(stakeEntry, totalReward), nil
+}
+
+func (k Keeper) updateDelegatorsReward(ctx sdk.Context, provider sdk.AccAddress, delegations []dualstakingtypes.Delegation, chainID string, totalReward math.Int) error {
+	stakeEntry, found, _ := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, chainID, provider)
```
Comment (Yaroms) [inline x/pairing/keeper/msg_server_relay_payment.go:449]: * we already have the stake entry from previous calls, why again?
* use the epoch of the payment, dont get the current stake

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 5]
```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+	if err != nil {
+		return utils.LavaFormatError("cannot update delegators reward map", err)
+	}
+
+	return nil
+}
+
+// calculateProviderDelegatorsRewards returns the provider reward amount and updates the delegator reward map
+func (k Keeper) calculateProviderDelegatorsRewards(ctx sdk.Context, provider sdk.AccAddress, chainID string, totalReward math.Int) (math.Int, error) {
+	stakeEntry, found, _ := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, chainID, provider)
```
Comment (Yaroms) [inline x/pairing/keeper/msg_server_relay_payment.go:427]: get the stake for the appropriate epoch

File: x/dualstaking/keeper/delegator_reward.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	return totalDelegations, totalDelegations.Add(stakeEntry.Stake.Amount)
+}
+
+// CalcDelegatorReward calculates a single delegator reward according to its delegation
+// delegatorReward = (delegatorsReward * delegatorStake) / totalDelegations
+func (k Keeper) CalcDelegatorReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int, delegation types.Delegation) math.Int {
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+
+	delegatorsReward := k.CalcDelegatorsReward(stakeEntry, totalReward)
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegator_reward.go:95]: this method is called in a for loop but always has the same result, call it before the loop once

File: x/dualstaking/keeper/delegator_reward.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+// CalcEffectiveDelegationsAndStake calculates the effective stake and effective delegations (for delegator rewards calculations)
+// effectiveDelegations = totalDelegations
+// effective stake = effectiveDelegations + providerStake
+func (k Keeper) CalcEffectiveDelegationsAndStake(stakeEntry epochstoragetypes.StakeEntry) (effectiveDelegations math.Int, effectiveStake math.Int) {
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	return totalDelegations, totalDelegations.Add(stakeEntry.Stake.Amount)
+}
+
+// CalcDelegatorReward calculates a single delegator reward according to its delegation
+// delegatorReward = (delegatorsReward * delegatorStake) / totalDelegations
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegator_reward.go:91]: for calrity:
// delegatorReward = delegatorsReward * (delegatorStake / totalDelegations) = (delegatorsReward * delegatorStake) / totalDelegations

File: x/pairing/keeper/helpers_test.go [hunk 8]
```diff
@@ -84,6 +86,14 @@ func (ts *tester) addProviderExtra(
 	return nil
 }
 
+// addDelegators adds CONSUMER accounts without buying a subscription to be used as delegators
+func (ts *tester) addDelegators(count int) {
```
Comment (Yaroms) [inline x/pairing/keeper/helpers_test.go:90]: why do we need this?
use AddAcounts instead

File: x/dualstaking/keeper/delegator_reward.go (hunk 7 above)
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/aaed62b60677647104a3761f38c86c333d59a3a3

File: x/dualstaking/keeper/delegator_reward.go (hunk 2 above)
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/aaed62b60677647104a3761f38c86c333d59a3a3

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 1 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go:437]: done. see https://github.com/lavanet/lava/commit/aaed62b60677647104a3761f38c86c333d59a3a3

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 4 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: done. see https://github.com/lavanet/lava/commit/4734092858e3f63ac104d340e6ed56a308a25f3c

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 5 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: done. see https://github.com/lavanet/lava/commit/4734092858e3f63ac104d340e6ed56a308a25f3c

File: x/pairing/keeper/helpers_test.go (hunk 8 above)
Comment (oren-lava) [reply x/pairing/keeper/helpers_test.go]: done. see https://github.com/lavanet/lava/commit/4985b44d36d626bb5dcb72cd8a1fe821e723a21a

File: x/dualstaking/keeper/delegator_reward.go (hunk 6 above)
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/4985b44d36d626bb5dcb72cd8a1fe821e723a21a

File: x/pairing/keeper/delegator_rewards_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/lavanet/lava/utils/slices"
+	dualstakingtypes "github.com/lavanet/lava/x/dualstaking/types"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	"github.com/lavanet/lava/x/pairing/types"
+	"github.com/stretchr/testify/require"
+)
+
+// TestProviderDelegatorsRewards tests that the provider's reward (considering delegations) is as expected
+// Also, it checks that the delegator reward map is updated as expected
+func TestProviderDelegatorsRewards(t *testing.T) {
```
Comment (orenl-lava) [inline x/pairing/keeper/delegator_rewards_test.go:19]: this needs to be repeated with multiple combinations:
(would cover the test explicit test cases in the next two functions)

commission = 0, 47, 100  (the "47" chosen to be unlikely to nicely divide the effective delegation)
limit = 0, larger than total delegation, smaller than total delegation
delegation ratios (between two delegators):   0/100, 10/90, 50/50  (the 14/86 is unlikely to nicely divide the total delegations)

so that's a matrix of 3x3x3 combinations to test. 
it will check edge cases, including the case of unclaimed leftovers that need to go to the provider.

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 10]
```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+// distributeRewards is the main function for reward distribution for providers and delegators
+func (k Keeper) distributeRewards(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, block uint64, totalReward math.Int) error {
+	epoch, _, err := k.epochStorageKeeper.GetEpochStartForBlock(ctx, block)
```
Comment (orenl-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:397]: why not use `block` directly?

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 11]
```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+	if err != nil {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", err,
```
Comment (orenl-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:399]: the first string repeats throughout this function. use a const?

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 12]
```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+	stakeEntry, found, stakeIndex := k.epochStorageKeeper.GetStakeEntryByAddressFromStorage(ctx, stakeStorage, providerAddr)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, stakeEntry, stakeIndex, totalReward)
```
Comment (orenl-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:420]: as per comment to the function called, this should be:
```
providerReward, delegatorsReward := ....
```

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 13]
```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+			utils.Attribute{Key: "reward", Value: providerRewardCoins},
+		)
+	}
+
+	delegations, err := k.dualStakingKeeper.GetProviderDelegators(ctx, providerAddr.String(), epoch)
+	if err != nil {
+		return utils.LavaFormatError("cannot get provider's delegators", err)
+	}
+
+	var relevantDelegations []dualstakingtypes.Delegation
```
Comment (orenl-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:440]: you can use `utils/slices`'s Filter() helper.

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 1 above)
Comment (orenl-lava) [reply x/pairing/keeper/msg_server_relay_payment.go:458]: I think this check is unnecessary (we do not verify all values/params all the time, do we?). 
But if the check is to remain, that it belongs where the value is used, i.e. `CalcProviderReward()`.
Thus, this function can be dropped altogether.

File: x/pairing/keeper/msg_server_relay_payment.go [hunk 14]
```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
...
+			utils.Attribute{Key: "chainID", Value: stakeEntry.Chain},
+			utils.Attribute{Key: "invalidCommission", Value: invalidCommission},
+		)
+	}
+
+	return k.dualStakingKeeper.CalcProviderReward(stakeEntry, totalReward), nil
+}
+
+// updateDelegatorsReward updates the delegator rewards map
+func (k Keeper) updateDelegatorsReward(ctx sdk.Context, stakeEntry epochstoragetypes.StakeEntry, delegations []dualstakingtypes.Delegation, totalReward math.Int) error {
```
Comment (orenl-lava) [inline x/pairing/keeper/msg_server_relay_payment.go:473]: I think here is unsuitable to do this work. 
It would make more sense for dualstaking to loop through the delegations and do this work, there. It already has all the know-how, just needs to know the totalReward. And then none of those internals of dualstaking needs to be exported to here. (For instance, how provider/chainID/delegator are encoded into keys is none of this pairing-module's business).

File: x/pairing/types/expected_keepers.go [hunk 15]
```diff
@@ -86,3 +89,13 @@ type PlanKeeper interface {
 type DowntimeKeeper interface {
 	GetDowntimeFactor(ctx sdk.Context, epochStartBlock uint64) uint64
 }
+
+type DualStakingKeeper interface {
```
Comment (orenl-lava) [inline x/pairing/types/expected_keepers.go:93]: I believe we can avoid all these exports if the core functionality remains in dualstaking.

**Review by orenl-lava — CHANGES_REQUESTED** [review]

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 10 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go:397]: because when I fetch stakeStorage, the function constructs a key that is dependent on the block value. Since stakeStorage objects are saved with keys that always contain blocks of epoch starting points, I need to get the start of the epoch

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 1 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: dropped the check. see https://github.com/lavanet/lava/commit/c23f6d56950de9bd0d143d7fb62471402a5354e2

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 11 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: done. see https://github.com/lavanet/lava/commit/c23f6d56950de9bd0d143d7fb62471402a5354e2

File: proto/lavanet/lava/dualstaking/delegator_reward.proto [hunk 16]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package lavanet.lava.dualstaking;
+
+option go_package = "github.com/lavanet/lava/x/dualstaking/types";
+
+import "gogoproto/gogo.proto";
+import "cosmos/base/v1beta1/coin.proto";
+
+message DelegatorReward {
+    string index = 1; // index = provider + chainID + delegator
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply proto/lavanet/lava/dualstaking/delegator_reward.proto:10]: I agree, but I've scaffolded this map like ignite would. Since I didn't want to deviate from the Cosmos scaffolded maps standard in our code (see EpochPayments, ProviderPaymentStorage, StakeStorage and more), I implemented this struct with an "index" field.

Should I change it?

File: x/dualstaking/keeper/delegator_reward.go [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+// CalcDelegatorsReward calculates the total amount of rewards for all delegators
+// delegatorsReward = totalReward - providerReward
+func (k Keeper) CalcDelegatorsReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
+	return totalReward.Sub(k.CalcProviderReward(stakeEntry, totalReward))
+}
+
+// CalcDelegationsSum calculates the delegations' sum
+// delegations sum = totalDelegations + providerStake
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go:89]: done. see https://github.com/lavanet/lava/commit/f9c49a16a43c79f77ffae7432d69b349a014f91a

File: x/dualstaking/keeper/delegator_reward.go [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	delegationsSum := k.CalcDelegationsSum(stakeEntry)
+
+	providerRewardPercentage := totalDelegations.MulRaw(int64(delegationCommission / 100)).Add(providerStake).Quo(delegationsSum)
+	return providerRewardPercentage.Mul(totalReward)
+}
+
+// CalcDelegatorsReward calculates the total amount of rewards for all delegators
+// delegatorsReward = totalReward - providerReward
+func (k Keeper) CalcDelegatorsReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/f9c49a16a43c79f77ffae7432d69b349a014f91a

File: x/dualstaking/keeper/delegator_reward.go [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+		var val types.DelegatorReward
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		list = append(list, val)
+	}
+
+	return
+}
+
+// CalcProviderReward calculates the provider reward considering delegations
+// providerReward = totalReward * ((totalDelegations*commission + providerStake) / delegationsSum)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go:75]: done. see https://github.com/lavanet/lava/commit/f9c49a16a43c79f77ffae7432d69b349a014f91a

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 12 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: outdated. There is no such function.
Also, CalcRewards was implemented here: https://github.com/lavanet/lava/commit/f9c49a16a43c79f77ffae7432d69b349a014f91a

File: x/dualstaking/keeper/delegator_reward.go [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+func (k Keeper) CalcProviderReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
+	providerStake := stakeEntry.Stake.Amount
+	delegationCommission := stakeEntry.DelegateCommission
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	delegationsSum := k.CalcDelegationsSum(stakeEntry)
+
+	providerRewardPercentage := totalDelegations.MulRaw(int64(delegationCommission / 100)).Add(providerStake).Quo(delegationsSum)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/20ebf36ed2735bb329fa75fdcca533104147aec2
As discussed offline, I won't add another field to stakeEntry

File: x/dualstaking/keeper/delegator_reward.go [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+// CalcDelegationsSum calculates the delegations' sum
+// delegations sum = totalDelegations + providerStake
+func (k Keeper) CalcDelegationsSum(stakeEntry epochstoragetypes.StakeEntry) math.Int {
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	return totalDelegations.Add(stakeEntry.Stake.Amount)
+}
+
+// CalcDelegatorReward calculates a single delegator reward according to its delegation
+// delegatorReward = delegatorsReward * (delegatorStake / totalDelegations) = (delegatorsReward * delegatorStake) / totalDelegations
+func (k Keeper) CalcDelegatorReward(delegatorsReward math.Int, totalDelegations math.Int, delegation types.Delegation) math.Int {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go:94]: done. see https://github.com/lavanet/lava/commit/20ebf36ed2735bb329fa75fdcca533104147aec2

File: proto/lavanet/lava/dualstaking/delegator_reward.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply proto/lavanet/lava/dualstaking/delegator_reward.proto]: instead of keeping index, I'm keeping its components. See https://github.com/lavanet/lava/commit/59a09b37fc751c2632666683bcd7d068adae645d

File: x/dualstaking/keeper/delegator_reward.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: I did a mistake in the last commit (I divided by totalDelegations and no delegationsSum). see fix here: https://github.com/lavanet/lava/commit/59a09b37fc751c2632666683bcd7d068adae645d

File: x/dualstaking/keeper/grpc_query_delegator_providers.go [hunk 22]
```diff
@@ -45,7 +45,6 @@ func (k Keeper) DelegatorProviders(goCtx context.Context, req *types.QueryDelega
 				)
 				continue
 			}
-			delegations = append(delegations, delegation)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/grpc_query_delegator_providers.go]: no. fixed here: https://github.com/lavanet/lava/commit/59a09b37fc751c2632666683bcd7d068adae645d

File: x/dualstaking/keeper/delegator_reward.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: done. see https://github.com/lavanet/lava/commit/bd381397394e90ce1e11a7dd46f016abf221d2ee

[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/dualstaking/keeper/delegator_reward.go]: nice.  but that doesn't include the effective stuff we discussed?
also, I suggest to complete the documentation here and explain also how delegations affect staking (even though it's done elsewhere), and how the provider/delegators/delegaor portions are calculated - so the intention is clear from the comment and not only from the code.

[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/dualstaking/keeper/delegator_reward.go]: ah. never mind. in next comment :)

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 13 above)
Comment (orenl-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: It's actually a `Map()` function, not filter. 
See #785

File: x/dualstaking/keeper/delegator_reward.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/dualstaking/keeper/delegator_reward.go]: unsure I see it .. wrong commit?

File: x/dualstaking/keeper/delegator_reward.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (orenl-lava) [reply x/dualstaking/keeper/delegator_reward.go:102]: nice!

File: proto/lavanet/lava/dualstaking/delegator_reward.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (Yaroms) [reply proto/lavanet/lava/dualstaking/delegator_reward.proto]: do we need the all of them, we have them when we fetch

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 14 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: done. see https://github.com/lavanet/lava/commit/520bc5955430cf6ec1125c970dc5e3ab3aca6560

File: x/pairing/keeper/msg_server_relay_payment.go (hunk 13 above)
Comment (oren-lava) [reply x/pairing/keeper/msg_server_relay_payment.go]: merged main and solved conflicts. Now using your implementation. see https://github.com/lavanet/lava/commit/8261db8d95f5cc75be14d2d0ff3bda802b5f7eee

File: x/dualstaking/keeper/delegator_reward.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/keeper/delegator_reward.go]: see here: https://github.com/lavanet/lava/commit/bd381397394e90ce1e11a7dd46f016abf221d2ee

File: proto/lavanet/lava/dualstaking/delegator_reward.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (oren-lava) [reply proto/lavanet/lava/dualstaking/delegator_reward.proto]: we need them to keep the delegator rewards data in the genesis state

File: x/pairing/keeper/delegator_rewards_test.go (hunk 9 above)
Comment (oren-lava) [reply x/pairing/keeper/delegator_rewards_test.go:19]: done. see https://github.com/lavanet/lava/commit/57489eca55ecc1b76b193a712bd87bb88e84f30b

Comment (orenl-lava) [reply x/pairing/keeper/delegator_rewards_test.go:19]: nice. (I was thinking of a 3-deep nested loops, of 3 cases each, but this works just as well.)

**Review by orenl-lava — DISMISSED** [review]

**Review by Yaroms — APPROVED** [review]
